File: core/src/main/java/org/dreamexposure/discal/core/object/calendar/CalendarData.java
package org.dreamexposure.discal.core.object.calendar;

import org.json.JSONObject;

import discord4j.common.util.Snowflake;

/**
 * Created by Nova Fox on 11/10/17.
 * Website: www.cloudcraftgaming.com
 * For Project: DisCal-Discord-Bot
 */
public class CalendarData {
    public static CalendarData fromJson(final JSONObject json) {
        final Snowflake guildId = Snowflake.of(json.getString("guild_id"));
        final int calendarNumber = json.getInt("calendar_number");

        final String calendarId = json.getString("calendar_id");
        final String calendarAddress = json.getString("calendar_address");
        final boolean external = json.getBoolean("external");
        final int credentialId = json.optInt("credential_id", 0);

        return new CalendarData(guildId, calendarNumber, calendarId, calendarAddress, external, credentialId);
    }

    public static CalendarData fromData(final Snowflake gId, final int calNum, final String calId,
                                        final String calAddr, final boolean ext, final int credId) {
        return new CalendarData(gId, calNum, calId, calAddr, ext, credId);
    }

    public static CalendarData empty() {
        return new CalendarData(Snowflake.of(0), 0, "primary", "primary", false, 0);
    }

    private final Snowflake guildId;
    private final int calendarNumber;

    private final String calendarId;
    private final String calendarAddress;

    private final boolean external;

    private final int credentialId;

    private CalendarData(final Snowflake guildId, final int calendarNumber, final String calendarId,
                         final String calendarAddress, final boolean external, final int credentialId) {
        this.guildId = guildId;
        this.calendarNumber = calendarNumber;
        this.calendarId = calendarId;
        this.calendarAddress = calendarAddress;
        this.external = external;
        this.credentialId = credentialId;
    }

    //Getters
    public Snowflake getGuildId() {
        return this.guildId;
    }

    public int getCalendarNumber() {
        return this.calendarNumber;
    }

    public String getCalendarId() {
        return this.calendarId;
    }

    public String getCalendarAddress() {
        return this.calendarAddress;
    }

    public boolean isExternal() {
        return this.external;
    }

    public int getCredentialId() {
        return this.credentialId;
    }

    public JSONObject toJson() {
        final JSONObject json = new JSONObject();

        json.put("guild_id", this.guildId.asString());
        json.put("calendar_number", this.calendarNumber);
        json.put("calendar_id", this.calendarId);
        json.put("calendar_address", this.calendarAddress);
        json.put("external", this.external);
        json.put("credential_id", this.credentialId);

        return json;
    }
}

File: core/src/main/kotlin/org/dreamexposure/discal/core/object/calendar/CalendarData.kt
package org.dreamexposure.discal.core.`object`.calendar

import discord4j.common.util.Snowflake
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.dreamexposure.discal.core.enums.calendar.CalendarHost
import org.dreamexposure.discal.core.serializers.SnowflakeAsStringSerializer
import org.dreamexposure.novautils.crypto.KeyGenerator
import java.time.Instant

@Serializable
data class CalendarData(
        @Serializable(with = SnowflakeAsStringSerializer::class)
        @SerialName("guild_id")
        val guildId: Snowflake = Snowflake.of(0),
        @SerialName("calendar_number")
        val calendarNumber: Int = 1,
        val host: CalendarHost,
        @SerialName("calendar_id")
        val calendarId: String = "primary",
        @SerialName("calendar_address")
        val calendarAddress: String = "primary",
        val external: Boolean = false,

        //secure values that should not be serialized
        @Transient
        val credentialId: Int = 0,
        @Transient
        var privateKey: String = KeyGenerator.csRandomAlphaNumericString(16),
        @Transient
        var encryptedAccessToken: String = "N/a",
        @Transient
        var encryptedRefreshToken: String = "N/a",
        @Transient
        var expiresAt: Instant = Instant.now()
) {
    constructor(guildId: Snowflake, calendarNumber: Int, host: CalendarHost, calendarId: String,
                calendarAddress: String, credentialId: Int) :
            this(guildId, calendarNumber, host, calendarId, calendarAddress, false, credentialId)

    companion object {
        @JvmStatic
        fun empty(guildId: Snowflake, host: CalendarHost) = CalendarData(guildId, host = host)

        fun emptyExternal(guildId: Snowflake, host: CalendarHost) = CalendarData(guildId, external = true, host = host)
    }

    fun expired() = Instant.now().isAfter(expiresAt)
}
